release/internal/pinnedversion/pinnedversion.go

```diff
@@ -47,167 +48,222 @@ var noImageComponents = []string{
 	"networking-calico",
 }
 
-type OperatorConfig struct {
-	// Dir is the directory to clone the operator repository.
-	Dir string
-
-	// Branch is the repository for the operator
-	Branch string
+type PinnedVersions interface {
+	GenerateFile() (version.Data, error)
+	ManagerOptions() ([]calico.Option, error)
+}
 
-	// Registry is the registry for Tigera operator
+type OperatorConfig struct {
+	Dir      string
+	Branch   string
 	Registry string
-
-	// Image is the image for Tigera operator
-	Image string
+	Image    string
 }
 
-func (c OperatorConfig) GitVersion() version.Version {
-	previousTag, err := command.GitVersion(c.Dir, true)
+func (c OperatorConfig) GitVersion() (string, error) {
+	tag, err := command.GitVersion(c.Dir, true)
 	if err != nil {
-		logrus.WithError(err).Fatal("Failed to determine latest git version")
+		logrus.WithError(err).Error("Failed to determine operator git version")
+		return "", err
 	}
-	logrus.WithField("out", previousTag).Info("Current git describe")
-	return version.New(previousTag)
+	logrus.WithField("out", tag).Info("Current git describe")
+	return tag, nil
 }
 
 func (c OperatorConfig) GitBranch() (string, error) {
 	return command.GitInDir(c.Dir, "rev-parse", "--abbrev-ref", "HEAD")
 }
 
-// Config represents the configuration needed to generate the pinned version file.
-type Config struct {
-	// RootDir is the root directory of the repository.
-	RootDir string
-
-	// ReleaseBranchPrefix is the prefix for the release branch.
-	ReleaseBranchPrefix string
-
-	// Operator is the configuration for the operator.
-	Operator OperatorConfig
+// PinnedVersion represents an entry in pinned version file.
+type PinnedVersion struct {
+	Title          string                        `yaml:"title"`
+	ManifestURL    string                        `yaml:"manifest_url"`
+	ReleaseName    string                        `yaml:"release_name,omitempty"`
+	Note           string                        `yaml:"note"`
+	Hash           string                        `yaml:"full_hash"`
+	TigeraOperator registry.Component            `yaml:"tigera-operator"`
+	Components     map[string]registry.Component `yaml:"components"`
 }
 
-// PinnedVersionData represents the data needed to generate the pinned version file from the template.
-type PinnedVersionData struct {
-	// ReleaseName is the name of the release.
-	ReleaseName string
+func NewCalicoVersionData(productVersion, operatorVersion string) *CalicoPinnedVersionData {
+	return &CalicoPinnedVersionData{
+		calico:   version.New(productVersion),
```

Comment:
It seems a little odd to me to parse these into `Version` structs, and then throughout the `CalicoPinnedVersionData` implementation we exclusively call `FormattedString` to convert back into a string. Is there a reason?

Reply:
I'm actually not sure what the purpose of this structure - `CalicoPinnedVersionData` - is.

I originally thought it was going to implement the `PinnedVersions` interface, but looks like that is actually done by `CalicoPinnedVersions` instead.

This struct seems to effectively take two strings and then provide methods to access those strings, which seems like it might be overcomplicated rather than just passing the strings around where needed. Unless there's an interface that this is intended to satsify with an alternative implementation?

Basically, I think the code is missing comments justifying why this construct should exist.

Reply:
I have moved this struct to `version` package. It implements `version.Data` interface